### PR TITLE
Refactored CMSIS-NN kernel additions as an example

### PR DESCRIPTION
**Pull request by petewarden**
Thanks for https://github.com/tensorflow/tensorflow/pull/27274 ! I wanted to give detailed feedback on it, and I thought the best approach might be a pull request to your change, so I can show you concretely my suggestions.

First, this is great, thanks again for putting this together, I'm excited to have CMSIS NN kernels making their way into TF Lite. This is an important step, and there's not much documentation available on adding specialized kernel implementations, so I wanted to go through the PR in detail to show what we're thinking of.

We're hoping that the "TAGS" method will be the main way to specify what optimized implementations are requested. The advantage of this is that if a word is included as a tag, then any implementations in subfolders (for example micro/kernels/cmsis-nn/depthwise_conv.cc) will replace the top-level reference versions (micro/kernels/depthwise_conv.cc) automatically.

The original PR used a separate KERNEL_LIB argument to do something similar, but this meant the makefile itself had to do a lot of the work of replacing implementations. I switched over to specifying TAGS="cmsis-nn" instead, like this:

`make -f tensorflow/lite/experimental/micro/tools/make/Makefile TAGS=cmsis-nn TARGET=bluepill test_depthwise_conv_test`

This meant I could remove a lot of the code in cmsis.inc. I did make one other change there too, removing the separate library step, and including the CMSIS NN .c files directly in the MicroLite source list. That removes the need to add library includes elsewhere in the code, but if you have a strong preference towards keeping a separate .a let me know and we can figure out a better approach.

Otherwise this looks awesome, so let me know what you think about this approach, and I'm keen to chat in person when we get a chance.

### CODE REVIEW

**Comment by freddan80** [comment]
Thanks for the nice example and feedback!

I like the solution using tags, but I couldn't find a good solution for microarchitecture-specific implementation. For example, if M0 has an optimized CMSIS-NN implementation for Fully Connected, but not for Depthwise Conv, while an M4 has optimized CMSIS-NN implementations for both. Then I'd like to use the 'cmsis-nn' tag for both architectures. However, for M0 I'd like to use the Depthwise Conv (TFLu) ref kernel. So somehow I want to build include cmsis-nn/fully_connected.cc, but exclude cmsis-nn/depthwise_conv.cc.

Let me investigate some more on how to do that with tags. Or do you know a solution already?

Regarding the lib. I was thinking it would be good to have the possibility to add one (or more) 3rd party lib, possibly closed source. A hypothetical example: We have a HW with an M0 which has an optimized CMSIS-NN implementation for Fully Connected, plus a DSP which has an optimized implementation for Depthwise Conv. For Softmax, I'd like to fall back on the ref kernel. I build it like:

make -f tensorflow/lite/experimental/micro/tools/make/Makefile TAGS=cmsis-nn my-dsp TARGET=my-hw test

These are my initial thoughts, let's discuss it more in detail in a chat/call.

**Comment by petewarden** [comment]
> I like the solution using tags, but I couldn't find a good solution for microarchitecture-specific implementation. For example, if M0 has an optimized CMSIS-NN implementation for Fully Connected, but not for Depthwise Conv, while an M4 has optimized CMSIS-NN implementations for both. Then I'd like to use the 'cmsis-nn' tag for both architectures. However, for M0 I'd like to use the Depthwise Conv (TFLu) ref kernel. So somehow I want to build include cmsis-nn/fully_connected.cc, but exclude cmsis-nn/depthwise_conv.cc.

That makes sense as a use case, and isn't easy to handle just with the tags system. I've ended up having a specialized version of depthwise conv that just calls back to the reference implementation for parameter sizes that it isn't optimized for:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/experimental/micro/kernels/portable_optimized/depthwise_conv.cc

So one option is having a kernel.cc file for every op that CMSIS-NN supports, but then in the code using #ifdef's to call the base reference implementation for any architectures that don't have optimized versions. That's not very elegant of course, so let me know what you think.

> Regarding the lib. I was thinking it would be good to have the possibility to add one (or more) 3rd party lib, possibly closed source. A hypothetical example: We have a HW with an M0 which has an optimized CMSIS-NN implementation for Fully Connected, plus a DSP which has an optimized implementation for Depthwise Conv. For Softmax, I'd like to fall back on the ref kernel. I build it like:

That definitely also makes sense as an important use case. The way I'm hoping that would work would be that we'd have stub kernels/<op>.cc source files that include a prebuilt library's headers, call into the underlying implementation, and then add the library path into LDOPTS in a Makefile.inc. Does that make sense?

**Comment by freddan80** [comment]
Thanks for the comments!

> That makes sense as a use case, and isn't easy to handle just with the tags system. I've ended up having a specialized version of depthwise conv that just calls back to the reference implementation for parameter sizes that it isn't optimized for:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/experimental/micro/kernels/portable_optimized/depthwise_conv.cc
> So one option is having a kernel.cc file for every op that CMSIS-NN supports, but then in the code using #ifdef's to call the base reference implementation for any architectures that don't have optimized versions. That's not very elegant of course, so let me know what you think.

I wanted to avoid the #ifdef's, but we might be able to do that somehow. Let me think about it. 

However, there'll be some code duplication with that approach, for example EvalFloat() exist in:

/micro/kernels/portable_optimized/depthwise_conv.cc
/micro/kernels/depthwise_conv.cc

With could be costly in maintenance, especially if there are many 3rd party optimized kernels. Or perhaps I have overlooked something.

One potential solution to that; What if the ref kernels are always compiled into the microlite lib. And the specialized implementations overrides (build time) only the registration of Invoke() etc. to the AllOpsResolver..? That would mean that the specialized implementations can fall back on Eval(), EvalFloat() etc. of the reference kernel. 

I'll make a prototype on Monday if the idea sounds reasonable.

**Comment by freddan80** [comment]
> That definitely also makes sense as an important use case. The way I'm hoping that would work would be that we'd have stub kernels/.cc source files that include a prebuilt library's headers, call into the underlying implementation, and then add the library path into LDOPTS in a Makefile.inc. Does that make sense?

Absolutely! So if I understand it correctly, in my example above, we'd then have:

kernels/softmax.cc
kernels/fully_connected.cc ---> overridden
kernels/depthwise_conv.cc ---> overridden
kernels/cmsis-nn/fully_connected.cc
kernels/cmsis-nn/depthwise_conv.cc ---> overridden
kernels/my-dsp/depthwise_conv.cc

To include both cmsis-nn and my-dsp implementations, the following is done:

make -f tensorflow/lite/experimental/micro/tools/make/Makefile TAGS=cmsis-nn my-dsp TARGET=my-hw test

in combination with a file 'my-dsp.inc' that includes:

LDOPTS=L<path to lib>/the-lib.a

Something like that..?
